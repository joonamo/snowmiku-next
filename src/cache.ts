import { S3Client, GetObjectCommand, PutObjectCommand } from '@aws-sdk/client-s3'

const Bucket = String(process.env.S3_BUCKET)
const client = new S3Client({
  credentials: {
    accessKeyId: String(process.env.S3_KEY),
    secretAccessKey: String(process.env.S3_SECRET)
  },
  endpoint: String(process.env.S3_URL),
  region: 'auto'
})

export interface CachedObject<T> {
  validUntil: number
  data: T
}

export const getCached = async <T>(key: string): Promise<T | undefined> => {
  const cacheResult = await client.send(new GetObjectCommand({
    Bucket,
    Key: key
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  })).catch((e: any) => {
    if (e.Code === 'NoSuchKey') {
      return undefined
    } else {
      throw e
    }
  })

  const cachedString = await cacheResult?.Body?.transformToString()

  if (!cachedString) {
    return undefined
  }

  const cachedObject: CachedObject<T> = JSON.parse(cachedString)
  if (new Date().getTime() < cachedObject.validUntil) {
    console.log(`cache hit "${key}"`)
    return cachedObject.data
  }

  return undefined
}

export const storeCache = async (key: string, data: unknown, lifetimeMillis: number) => {
  const Expires = new Date(new Date().getTime() + lifetimeMillis)
  
  await client.send(new PutObjectCommand({
    Bucket,
    Key: key,
    Body: JSON.stringify({ data, validUntil: Expires.getTime() }),
    Expires
  }))
}
