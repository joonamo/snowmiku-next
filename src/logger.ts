export const logInfo = (message: string, data: unknown) =>
  console.log(JSON.stringify({ message, data }))
export const logError = (message: string, data: unknown) =>
  console.error(JSON.stringify({ message, data }))
