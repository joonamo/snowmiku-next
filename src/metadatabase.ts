type YearMetadata = {
  winner?: string
  finalists?: string[]
  officialPage?: string
  theme?: string
  themeTranslated?: string
}

export const defaultTheme = '雪ミク'
export const defaultThemeTranslated = 'Snow Miku'

// Look at this fancy database!
export const metadatabase: Record<string, YearMetadata> = {
  '2024': {
    winner: 'H6zW',
    finalists: ['lixB', 'J9L4', 'hSdP', '9nud', 'H6zW', 'R29y'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2024snowmiku_p7hemqu6/',
    theme: 'ごちそう',
    themeTranslated: 'Feast',
  },
  '2023': {
    winner: 'sMw9',
    finalists: ['f23G', 'MVrT', 'mUlf', 'sMw9', 'p0xi', 'iQlv'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2023snowmiku_s5gnpkaj/',
    theme: '空模様',
    themeTranslated: 'Patterns of the Sky',
  },
  '2022': {
    winner: '-URY',
    finalists: ['NE6v', '-URY', 'IzMb', 'CPH6', 'nay8', '2m3q'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2022snowmiku_pb4vp5/',
    theme: '海',
    themeTranslated: 'Ocean',
  },
  '2021': {
    winner: 'CySS',
    finalists: ['CySS', 'ibL9', 'oJbN', '4hY9', 'bjDz', '6BM0'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2021snowmiku',
    theme: 'イルミネーション',
    themeTranslated: 'Illumination',
  },
  '2020': {
    winner: 'HnGM',
    finalists: ['HnGM', 'YxrR', 'saxQ', 'sWBP', 'lO2m', 'P9Kf'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2020snowmiku/',
    theme: '楽器',
    themeTranslated: 'Musical Instruments',
  },
  '2019': {
    winner: '70Dn',
    finalists: ['70Dn', '1oP2', '579f', 'e6cX', 'Mne6', 'liso'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2019snowmiku/',
    theme: 'プリンセス',
    themeTranslated: 'Princess',
  },
  '2018': {
    winner: 'u9XO',
    finalists: ['u9XO', 'faiV', '5r3E', 'qyP6', '_ytg', 'xBrM'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2018snowmiku',
    theme: 'どうぶつ',
    themeTranslated: 'Animals',
  },
  '2017': {
    winner: 'HKTc',
    finalists: ['HKTc', 'v0kt', '1WSt', 'S5v3', 'muF1', 'PYTJ'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2017snowmiku',
    theme: '星空/星座',
    themeTranslated: 'Starry Sky/Constellations',
  },
  '2016': {
    winner: 'uDla',
    finalists: ['uDla', 'tfBD', 'cAEy', 'XXfw'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2016snowmiku',
    theme: 'ウインタースポーツ',
    themeTranslated: 'Winter Sports',
  },
  '2015': {
    winner: 'w8W1',
    finalists: ['w8W1'],
    officialPage: 'https://piapro.jp/pages/official_collabo/2015snowmiku/',
    theme: '植物',
    themeTranslated: 'Plants',
  },
  '2014': {
    winner: 'xy8T',
    finalists: ['xy8T'],
    officialPage: 'https://piapro.jp/static/?view=2014yukimiku',
    theme: '魔法少女',
    themeTranslated: 'Magical Girl',
  },
  '2013': {
    winner: 'XSGC',
    finalists: ['XSGC'],
    officialPage: 'https://piapro.jp/static/?view=2013yukimiku',
    theme: 'お菓子',
    themeTranslated: 'Sweets',
  },
  '2012': {
    winner: 'Broe',
    finalists: ['Broe'],
    officialPage: 'https://piapro.jp/static/?view=2012yukimiku',
    theme: '宝石',
    themeTranslated: 'Jewels'
  },
}
