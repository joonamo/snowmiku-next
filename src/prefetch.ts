const baseurl =
  process.env.NODE_ENV === 'development'
    ? 'http://localhost:3000'
    : process.env.VERCEL_URL && `https://${process.env.VERCEL_URL}`

export const prefetchPage = (year: string, orderTag: string, page: number) => {
  if (baseurl) {
    const apiUrl = `${baseurl}/api/${orderTag}?year=${year}&page=${page}&noprefetch=${true}`
    console.log(`Prefetching ${baseurl}`)
    void fetch(apiUrl)
  }
}
