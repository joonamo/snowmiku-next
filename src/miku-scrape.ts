import JSSoup, { SoupTag } from 'jssoup'
import { getCached, storeCache } from './cache'

export interface MikuResult {
  name: string
  author: string
  link: string
  image: string
  authorIcon?: string
  piaproUrl?: string
}

export interface ResultsPage {
  results: MikuResult[]
  pageCount: number
}

export const latestTag = "sd"
export const popularTag = "cv"

export const getYearTag = (year: string) => {
  const tens = year[year.length - 2]
  const ones = year[year.length - 1]
  return `%EF%BC%92%EF%BC%90%EF%BC%9${tens}%EF%BC%9${ones}`
}

const imageRegex = /url\(\/\/(.*)0150_0150(\..*)\)/
const imageLink = (styleAttr: string) => {
  const re = imageRegex.exec(styleAttr)
  return `https://${re?.[1]}0740_0500${re?.[2]}`
}

export const paginatorRegex = /new Paginator\('_paginator', ([0-9]*)/
const pageCount = (page?: string) => Number((page && paginatorRegex.exec(page)?.[1]) ?? 1)

export const processPage = async (year: string, orderTag: string, page = 1): Promise<ResultsPage> => {    
  const cacheKey = `page-result/${year}/${orderTag}/${page}`
  const cached = await getCached<ResultsPage>(cacheKey)
  if (cached) {
    return cached
  }
  
  const yearTag = getYearTag(year)
  const piaproUrl = `https://piapro.jp/content_list/?view=image&tag=${yearTag}%E5%B9%B4%E9%9B%AA%E3%83%9F%E3%82%AF%E8%A1%A3%E8%A3%85&order=${orderTag}&page=${page}`
  
  console.log(piaproUrl)
  const mikuReq = await fetch(piaproUrl)
  console.log(`got response, status ${mikuReq.status} ${mikuReq.ok}`)
  const mikuHtml = await mikuReq.text()
  const soup = new JSSoup(mikuHtml)
  const images = soup.findAll('div', 'i_main')

  const results = images.map((item: SoupTag) => {
    const linkElem: SoupTag = item.find(undefined, 'i_image')
    return {
      name: item.find(undefined, 'thumb_over').text,
      author: item.find(undefined, 'i_title').text,
      authorIcon: item.find(undefined, 'i_icon')?.find('img').attrs['href'] ?? null,
      image: imageLink(linkElem.attrs['style']),
      link: `https://piapro.jp${linkElem.attrs['href']}`,
      piaproUrl
    }
  })
  
  const result = {
    pageCount: pageCount(mikuHtml),
    results
  }

  await storeCache(cacheKey, result, 5 * 60 * 1000)

  return result
}

export const getLatestYear = async (): Promise<number> => {
  const maxYear = new Date().getFullYear() + 1
  const promises = []
  for (let year = maxYear; year >= 2023; year--) {
    promises.push(processPage(year.toString(), latestTag))
  }

  const results = await Promise.all(promises)
  return maxYear - results.findIndex(result => result.results.length != 0)
}
