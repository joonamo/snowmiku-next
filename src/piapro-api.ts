import JSSoup, { SoupTag } from 'jssoup'
import { getCached, storeCache } from './cache'
import { logInfo } from './logger'
import { metadatabase } from './metadatabase'
import { randomUUID } from 'node:crypto'

export interface MikuResult {
  name: string
  author: string
  link: string
  image: string
  views: number
  postTime: string
  authorIcon?: string
  isFinalist?: boolean
  isWinner?: boolean
}

export interface ResultsPage {
  results: MikuResult[]
  piaproUrl: string
  pageCount: number
}

export const latestTag = 'sd'
export const popularTag = 'cv'

export const getYearTag = (year: string) => {
  const tens = year[year.length - 2]
  const ones = year[year.length - 1]
  return `%EF%BC%92%EF%BC%90%EF%BC%9${tens}%EF%BC%9${ones}`
}

const getId = (href?: string) => {
  if (!href) {
    return randomUUID()
  }

  const split = href.split('/')
  return split[split.length - 1]
}

// Deprecated
// const imageRegex = /url\(\/\/(.*)0250_0250(\..*)\)/
// const imageLink = (styleAttr: string) => {
//   const re = imageRegex.exec(styleAttr)
//   return `https://${re?.[1]}0740_0500${re?.[2]}`
// }

// const viewsRegex = /閲覧数:([\d,]+)/
// const getViews = (html: string) => {
//   const re = viewsRegex.exec(html)
//   return (re && parseInt(re[1].replace(',', ''), 10)) ?? 0
// }

// const postTimeRegex = /(\d\d\d\d\/\d\d\/\d\d \d\d:\d\d)/ // Beautiful
// const getPostTime = (html: string) => {
//   const re = postTimeRegex.exec(html)
//   return re?.[1] ?? ''
// }

export const paginatorRegex = /new Paginator\('_paginator', ([0-9]*)/
const pageCount = (page?: SoupTag) => {
  const pager = page?.find(undefined, 'pager_list')
  const last = pager.contents[pager.contents.length - 1]
  return parseInt(last.text, 10)
}

export const processPage = async (
  year: string,
  orderTag: string,
  page = 1,
): Promise<ResultsPage> => {
  const cacheKey = `page-result-v7/${year}/${orderTag}/${page}`
  const cached = await getCached<ResultsPage>(cacheKey)
  if (cached.data) {
    return cached.data
  }

  const yearTag = getYearTag(year)
  const piaproUrl = `https://piapro.jp/content_list/?view=image&tag=${yearTag}%E5%B9%B4%E9%9B%AA%E3%83%9F%E3%82%AF%E8%A1%A3%E8%A3%85&order=${orderTag}&page=${page}`

  logInfo(`Calling Piapro`, { piaproUrl })
  const mikuReq = await fetch(piaproUrl)
  console.log(`got response`, { status: mikuReq.status, ok: mikuReq.ok })
  const mikuHtml = await mikuReq.text()

  const processed = processHtml(mikuHtml, year)
  const result = { ...processed, piaproUrl }

  await storeCache(cacheKey, result, 5 * 60 * 1000)

  return result
}

export const processHtml = (mikuHtml: string, year: string) => {
  const soup = new JSSoup(mikuHtml)
  const images = soup.find(undefined, 'tmblist_list').contents
  const yearFinalists = metadatabase[year]

  const results = images.map((item: SoupTag): MikuResult => {
    const linkElem: SoupTag = item.find('a')
    const id = getId(linkElem.attrs['href'] as string)

    return {
      name: item.find(undefined, 'tmblist_list_title').text,
      author: item.find(undefined, 'tmblist_list_creator_txt').text,
      authorIcon:
        item.find(undefined, 'tmblist_list_creator_userimg')?.find('img').attrs['src'].replace('_0048.', '_0150.') ??
        null,
      image: item.find(undefined, 'tmblist_list_tmb_inner')?.find('img').attrs['src'].replace('0250_0250.', '0860_0600.'),
      link: `https://piapro.jp/t/${id}`,
      isFinalist: Boolean(yearFinalists?.finalists?.includes(id)),
      isWinner: yearFinalists?.winner === id,

      // Views and post time not available in current Piapro listing
      views: 0,
      postTime: '',
    }
  })

  const result = {
    pageCount: pageCount(soup),
    results,
  }

  return result
}

export const getLatestYear = async (): Promise<number> => {
  const cacheKey = 'latestYear'
  const cached = await getCached<{ latestYear: number }>(cacheKey, true)
  if (cached.data?.latestYear && !cached.stale) {
    return cached.data.latestYear
  }

  const maxYear = new Date().getFullYear() + 1
  let latestYear = cached.data?.latestYear
  if (maxYear === latestYear) {
    console.log('Stale cache, but valid year, accepting for 24h')
    await storeCache(cacheKey, { latestYear }, 24 * 60 * 60 * 1000)
  } else {
    const promises = []
    for (let year = maxYear; year >= 2024; year--) {
      promises.push(processPage(year.toString(), latestTag))
    }

    const results = await Promise.all(promises)
    latestYear = maxYear - results.findIndex((result) => result.results.length != 0)
    await storeCache(cacheKey, { latestYear }, 60 * 60 * 1000)
  }

  return latestYear
}
