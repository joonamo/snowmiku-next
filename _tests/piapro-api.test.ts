import { processHtml } from '../src/piapro-api'
import { expect, test } from 'vitest'
import { readFileSync } from 'node:fs'


test('Parses html', () => {
  const mikuHtml = readFileSync('./_tests/fixture/content_list.html').toString()
  const result = processHtml(mikuHtml, '2025')

  expect(result).toMatchSnapshot()
})