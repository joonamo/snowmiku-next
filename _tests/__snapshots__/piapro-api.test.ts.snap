// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parses html 1`] = `
{
  "pageCount": NaN,
  "results": [
    {
      "author": "user 1",
      "authorIcon": "//cdn.example.com/icon_u/873/111111_20240229160752_0072.jpg",
      "image": "//cdn.example.com/thumb_i/hf/aaaaa_20240229153653_0860_0600.png",
      "isFinalist": false,
      "isWinner": false,
      "link": "https://piapro.jp/t/AAAA",
      "name": "SNOW MIKU a",
      "postTime": "",
      "views": 0,
    },
    {
      "author": "user 2",
      "authorIcon": "//cdn.example.com/icon_u/845/22222_20240223120025_0072.jpg",
      "image": "//cdn.example.com/thumb_i/eh/bbbbbb_20240229130931_0860_0600.png",
      "isFinalist": false,
      "isWinner": false,
      "link": "https://piapro.jp/t/BBBB",
      "name": "雪ミク b",
      "postTime": "",
      "views": 0,
    },
    {
      "author": "user 3",
      "authorIcon": "//cdn.example.com/icon_u/449/33333_20240217035936_0072.jpg",
      "image": "//cdn.example.com/thumb_i/w4/ccccc_20240229063132_0860_0600.png",
      "isFinalist": false,
      "isWinner": false,
      "link": "https://piapro.jp/t/CCCC",
      "name": "スノーミク c",
      "postTime": "",
      "views": 0,
    },
  ],
}
`;
